app/models/grda_warehouse/auth_policies/policy_provider.rb

```diff
@@ -0,0 +1,127 @@
+  ###
+  # The legacy permission policy only checks the permissions on a user's roles and assumes the loaded record is visible
+  # to the user already. This is a dangerous assumption so it's required to confirm the record is authorized by
+  # setting legacy_implicitly_assume_authorized_access to true.
+  # Note this is only relevant for users where using_acls is false
+  ###
+  attr_accessor :legacy_implicitly_assume_authorized_access
+  # END_ACL
+
+  def initialize(user)
+    @user = user
+  end
+
+  def for_client(client_or_id)
+    if user.using_acls?
+      client_id = id_from_arg(client_or_id, GrdaWarehouse::Hud::Client)
+      for_client_using_acls(client_id)
+    else
+      # TODO: START_ACL remove after ACL migration is complete
+      legacy_user_role_policy
+      # END_ACL
+    end
+  end
+
+  def for_patient(patient)
+    for_client(patient.client_id)
+  end
+
+  def for_project(project_or_id)
+    if user.using_acls?
+      project_id = id_from_arg(project_or_id, GrdaWarehouse::Hud::Project)
+      for_project_using_acls(project_id)
+    else
+      # TODO: START_ACL remove after ACL migration is complete
+      legacy_user_role_policy
+      # END_ACL
+    end
+  end
+
+  protected
+
+  memoize def for_project_using_acls(project_id)
+    p_t = GrdaWarehouse::Hud::Project.arel_table
+    collection_ids = GrdaWarehouse::ProjectCollectionMember.where(project_id: project_id).pluck(:collection_id)
+
+    coc_codes = GrdaWarehouse::Hud::ProjectCoc.
+      joins(:project).
+      where(p_t[:id].eq(project_id)).
+      pluck(:coc_code)
+    collection_ids += Collection.for_coc_codes(coc_codes).pluck(:id) if coc_codes.any?
+
+    collection_ids += system_collection_ids(:data_sources)
+    GrdaWarehouse::AuthPolicies::CollectionPolicy.new(user: user, collection_ids: collection_ids)
+  end
+
+  memoize def for_client_using_acls(client_id)
+    c_t = GrdaWarehouse::Hud::Client.arel_table
+    gve_t = GrdaWarehouse::GroupViewableEntity.arel_table
+
+    # collections for the client's enrolled projects via HUD relationships. This is most common
+    collection_ids = GrdaWarehouse::ProjectCollectionMember.
+      joins(project: :clients).
+      where(c_t[:id].eq(client_id)).
+      pluck(:collection_id)
+
+    # collections for the client's enrolled projects using coc codes.
+    coc_codes = GrdaWarehouse::Hud::ProjectCoc.
+      joins(project: :clients).
+      where(c_t[:id].eq(client_id)).
+      pluck(:coc_code)
+    collection_ids += Collection.for_coc_codes(coc_codes).pluck(:id) if coc_codes.any?
+
+    # collections for the client's authoritative data source. Needed for clients records that do not have enrollments
+    collection_ids += GrdaWarehouse::DataSource.authoritative.not_hmis.
+      joins(:group_viewable_entities, :clients).
+      where(gve_t[:collection_id].not_eq(nil)).
+      where(c_t[:id].eq(client_id)).
+      pluck(gve_t[:collection_id])
+
+    collection_ids += system_collection_ids(:data_sources)
```

Comment:
Same as with projects, I _think_ this should move to `Collection`, and probably needs to handle the scenario where the entity assigned to the collection is an organization or data source.

Reply:
> Same as with projects, I think this should move to Collection

I could see an argument for putting the code to gather collections in a model. I'm a little reluctant to do that since there's so many auth methods scattered around, I find it hard to follow the the thread. Keeping the code internal to the class is a trade off, I find it to be more readable but it's definitely less re-usable. I guess I could go either way

> probably needs to handle the scenario where the entity assigned to the collection is an organization, project group, or data source.

That should be handled by ProjectCollectionMember up on line 74